### PR TITLE
AES CCM support

### PR DESCRIPTION
**Pull request by reaperhulk**
Depends on #3698 and #3699. WIP until those merge.

### CODE REVIEW

**Comment by alex** [comment]
This can be rebased now.

File: docs/hazmat/primitives/aead.rst [hunk 1]
```diff
@@ -77,3 +77,92 @@ also support providing integrity for associated data which is not encrypted.
...
+
+    .. method:: encrypt(nonce, data, associated_data, tag_length=16)
+
+        .. warning::
+
+            Reuse of a ``nonce`` with a given ``key`` compromises the security
+            of any message with that ``nonce`` and ``key`` pair.
+
+        Encrypts the ``data`` provided and authenticates the
+        ``associated_data``.  The output of this can be passed directly
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: It also authenticates the ``data``, please rephrase.

File: docs/hazmat/primitives/aead.rst [hunk 2]
```diff
@@ -77,3 +77,92 @@ also support providing integrity for associated data which is not encrypted.
...
+        to the ``decrypt`` method.
+
+        :param bytes nonce: A value of between 7 and 13 bytes. The maximum
+            length is determined by the length of the ciphertext you are
+            encrypting and must satisfy the condition:
+            ``len(data) < 2 ** (8 * (15 - len(nonce)))``
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: Is this right? This indicates that the longer the nonce the shorter the ciphertext must be?

File: docs/hazmat/primitives/aead.rst [hunk 3]
```diff
@@ -77,3 +77,92 @@ also support providing integrity for associated data which is not encrypted.
...
+            **NEVER REUSE A NONCE** with a key.
+        :param bytes data: The data to encrypt.
+        :param bytes associated_data: Additional data that should be
+            authenticated with the key, but does not need to be encrypted. Can
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: s/does not need to be/is not/

File: docs/hazmat/primitives/aead.rst [hunk 4]
```diff
@@ -77,3 +77,92 @@ also support providing integrity for associated data which is not encrypted.
...
+            be ``None``.
+        :param int tag_length: The length of the authentication tag. This
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: Is there a reason to support truncated tags?

File: docs/hazmat/primitives/aead.rst [hunk 5]
```diff
@@ -77,3 +77,92 @@ also support providing integrity for associated data which is not encrypted.
...
+        called encrypt with ``associated_data`` you must pass the same
+        ``associated_data`` in decrypt or the integrity check will fail.
+
+        :param bytes nonce: A value of between 7 and 13 bytes. This
+            is the same value used when you originally called encrypt.
+            **NEVER REUSE A NONCE** with a key.
+        :param bytes data: The data to decrypt (with tag appended).
+        :param bytes associated_data: Additional data to authenticate. Can be
+            ``None`` if none was passed during encryption.
+        :param int tag_length: The length of the authentication tag. This
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: This param isn't in the signature, but I'm opposed to it :-)

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 6]
```diff
@@ -11,8 +11,9 @@
 _DECRYPT = 0
 
 
-def _chacha20poly1305_setup(backend, key, nonce, tag, operation):
-    evp_cipher = backend._lib.EVP_get_cipherbyname(b"chacha20-poly1305")
+def _aead_setup(backend, cipher_name, key, nonce, tag, tag_len, operation):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: Please split these changes out into their own PR :-)

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 7]
```diff
@@ -1836,19 +1836,39 @@ def derive_scrypt(self, key_material, salt, length, n, r, p):
...
+    def aesccm_decrypt(self, key, nonce, data, associated_data, tag_length):
+        cipher_name = "aes-{0}-ccm".format(len(key) * 8).encode("ascii")
+        return aead._decrypt(
+            self, cipher_name, key, nonce, data, associated_data, tag_length
+        )
+
+    def aead_cipher_supported(self, cls):
+        # We do this ugly thing because circular import problems prevent
+        # importing aead so we could just do isinstance checks
+        if cls.__name__ == "ChaCha20Poly1305":
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Just importing the class in this function should work fine?

File: src/cryptography/hazmat/primitives/ciphers/aead.py [hunk 8]
```diff
@@ -12,7 +12,7 @@
 
 class ChaCha20Poly1305(object):
     def __init__(self, key):
-        if not backend.chacha20poly1305_supported():
+        if not backend.aead_cipher_supported(self.__class__):
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: prefer `type(self)`

File: src/cryptography/hazmat/primitives/ciphers/aead.py [hunk 9]
```diff
@@ -52,3 +52,67 @@ def _check_params(self, nonce, data, associated_data):
...
+
+class AESCCM(object):
+    def __init__(self, key):
+        utils._check_bytes("key", key)
+        if len(key) not in (16, 24, 32):
+            raise ValueError("AESCCM key must be 128, 192, or 256 bits.")
+
+        self._key = key
+
+        if not backend.aead_cipher_supported(self.__class__):
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: Same

File: src/cryptography/hazmat/primitives/ciphers/aead.py [hunk 10]
```diff
@@ -52,3 +52,67 @@ def _check_params(self, nonce, data, associated_data):
...
+    def decrypt(self, nonce, data, associated_data, tag_length=16):
+        if associated_data is None:
+            associated_data = b""
+
+        self._check_params(nonce, data, associated_data, tag_length)
+        return backend.aesccm_decrypt(
+            self._key, nonce, data, associated_data, tag_length
+        )
+
+    def _validate_l(self, nonce, data_len):
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: give this a clearer name please

File: src/cryptography/hazmat/primitives/ciphers/aead.py [hunk 11]
```diff
@@ -52,3 +52,67 @@ def _check_params(self, nonce, data, associated_data):
...
+        if not isinstance(tag_length, int):
+            raise TypeError("tag_length must be an integer")
+
+        if tag_length not in (4, 6, 8, 12, 14, 16):
+            raise ValueError("Invalid tag_length")
+
+        utils._check_bytes("nonce", nonce)
+        utils._check_bytes("data", data)
+        utils._check_bytes("associated_data", associated_data)
+        if len(nonce) < 7 or len(nonce) > 13:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: `if not 7 <= len(nonce) <= 13:` is slightly more readable

File: docs/hazmat/primitives/aead.rst (hunk 4 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/aead.rst]: CCM tags change based on length specified (rather than GCM's trivial truncation) and some specs don't use full length tags (including some TLS 1.2 cipher suites which hopefully nobody is implementing as well as IPsec, which mandates 8 and 16 byte tag support and says 12 is optional).

File: docs/hazmat/primitives/aead.rst (hunk 5 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/aead.rst]: Can't avoid it when supporting multiple tag lengths (which, as previously noted, I think we need to).

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 7 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Circular import because we import backend in cryptography.hazmat.primitives.ciphers.aead :(

File: docs/hazmat/primitives/aead.rst (hunk 1 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/aead.rst]: (Note to self: whatever we end up with on rephrasing we should also use it for chacha as this is the same text as the chacha docs use)

File: docs/hazmat/primitives/aead.rst (hunk 2 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/aead.rst]: Yes, the length of the ciphertext shortens the ciphertext because the length is encoded as part of the nonce. See https://tools.ietf.org/html/rfc3610#section-2

File: docs/hazmat/primitives/aead.rst (hunk 3 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/aead.rst]: (Same on porting new language to chacha)

File: src/cryptography/hazmat/primitives/ciphers/aead.py (hunk 8 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: Sure, what's the advantage for future reference? Beyond the general "any time you don't directly reference a dunder that's a win"?

File: src/cryptography/hazmat/primitives/ciphers/aead.py (hunk 10 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: In the RFC this is named `L` and the description is "Number of octets in length field". It represents the encoding of the length of the encrypted payload into bytes. `_validate_nonce_payload_length` maybe? I feel like `validate_l` is actually a good name given the fact that any debugging is going to necessitate a visit to the RFC.

**Comment by alex** [comment]
Rebase required :-)

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 12]
```diff
@@ -74,6 +86,9 @@ def _encrypt(backend, cipher_name, key, nonce, data, associated_data,
     ctx = _aead_setup(
         backend, cipher_name, key, nonce, None, tag_length, _ENCRYPT
     )
+    # CCM requires us to pass the length of the data before processing anything
+    if cipher_name[-3:] == b"ccm":
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: I don't like checking on this string. Let's think of a better API.

Can we just call `_set_length` for all AEADs?

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 7 above)
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Why does that prevent us from doing an import inside this function?

File: src/cryptography/hazmat/primitives/ciphers/aead.py (hunk 10 above)
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/aead.py]: How about `vaildate_lengths`, since it looks at `len(nonce)` and `len(data)`?

File: tests/hazmat/primitives/test_aead.py [hunk 13]
```diff
@@ -31,7 +35,7 @@ def test_chacha20poly1305_unsupported_on_older_openssl(backend):
 
 
 @pytest.mark.supported(
-    only_if=lambda backend: backend.chacha20poly1305_supported(),
+    only_if=lambda backend: backend.aead_cipher_supported(ChaCha20Poly1305),
```
Comment (alex) [inline tests/hazmat/primitives/test_aead.py]: If we're switching to an `aead_cipher_supported` API, can we split that out into a separate PR (lol)

File: src/cryptography/hazmat/backends/openssl/aead.py (hunk 12 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/aead.py]: Just tried, nope :(

**Comment by alex** [comment]
Ready for YAR (Yet Another Rebase)

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 14]
```diff
@@ -84,6 +103,10 @@ def _encrypt(backend, cipher_cls, key, nonce, data, associated_data,
     ctx = _aead_setup(
         backend, cipher_name, key, nonce, None, tag_length, _ENCRYPT
     )
+    # CCM requires us to pass the length of the data before processing anything
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: Move the import to the top of the function

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 15]
```diff
@@ -84,6 +103,10 @@ def _encrypt(backend, cipher_cls, key, nonce, data, associated_data,
...
+    from cryptography.hazmat.primitives.ciphers.aead import AESCCM
+    if cipher_cls is AESCCM:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: what happened if we called this for other aeads?

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 16]
```diff
@@ -111,12 +134,29 @@ def _decrypt(backend, cipher_cls, key, nonce, data, associated_data,
...
     _process_aad(backend, ctx, associated_data)
-    processed_data = _process_data(backend, ctx, data)
-    outlen = backend._ffi.new("int *")
-    res = backend._lib.EVP_CipherFinal_ex(ctx, backend._ffi.NULL, outlen)
-    if res == 0:
-        backend._consume_errors()
-        raise InvalidTag
+    # CCM has a different error path if the tag doesn't match. Errors are
+    # raised in Update and Final is irrelevant.
+    if cipher_name[-3:] == b"ccm":
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: no string typing

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 17]
```diff
@@ -13,11 +13,18 @@
 
 def _aead_cipher_name(cls, key_length):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: Ok, rather than pass the type here and doing all this weird `if key_lenght is None` stuff, let's pass an instance of the aead instead of the `cls`, then we can use the `key` attr.

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 18]
```diff
@@ -111,12 +134,29 @@ def _decrypt(backend, cipher_cls, key, nonce, data, associated_data,
...
-    res = backend._lib.EVP_CipherFinal_ex(ctx, backend._ffi.NULL, outlen)
-    if res == 0:
-        backend._consume_errors()
-        raise InvalidTag
+    # CCM has a different error path if the tag doesn't match. Errors are
+    # raised in Update and Final is irrelevant.
+    if cipher_name[-3:] == b"ccm":
+        outlen = backend._ffi.new("int *")
+        buf = backend._ffi.new("unsigned char[]", len(data))
+        res = backend._lib.EVP_CipherUpdate(ctx, buf, outlen, data, len(data))
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: Why can't this go through `_process_data`?

File: src/cryptography/hazmat/backends/openssl/aead.py (hunk 15 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/aead.py]: internal error out of openssl

Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py]: that's obnoxious. can you put in the comment that calling this on other AEAD's is an internal error

File: src/cryptography/hazmat/backends/openssl/aead.py (hunk 18 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/aead.py]: Because we need to check the error code and `_process_data` asserts on the error code and returns the buffer. Plus, calling EVP_CipherFinal returns zero with CCM (what a horribly overloaded API EVP is)
